File: soquete/TCP/cliente.ts
import * as net from 'net';

// Criar um cliente de socket
const client: net.Socket = net.createConnection({
  host: '127.0.0.1',
  port: 3001
});

// Lidar com mensagens do servidor
// Um buffer é uma área de memória temporária usada para armazenar dados durante a transferência ou processamento. 
// No Node.js, ele lida com dados binários de forma eficiente, atuando como ponte entre strings e dados binários em 
// operações de E/S e rede.
client.on('data', (mensagem: Buffer) => {
  console.log('Mensagem recebida do servidor:', mensagem.toString('utf8'));
});

// Lidar com a desconexão do servidor
client.on('close', () => {
  console.log('Conexão fechada');
});

// Enviar uma mensagem para o servidor
client.write('Olá, servidor!\r\n', () => {
  //client.end();
});

File: soquete/TCP/cliente.py
import socket

def tcp_client():
    # Configurações do servidor
    host = '127.0.0.1'
    port = 3001

    # Criar socket TCP
    client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    
    try:
        # Conectar ao servidor
        client_socket.connect((host, port))
        print(f"Conectado ao servidor {host}:{port}")

        # Enviar mensagem
        message = "Olá, servidor!\r\n"
        client_socket.sendall(message.encode('utf-8'))
        print(f"Mensagem enviada: {message.strip()}")

        # Receber resposta
        while True:
            data = client_socket.recv(1024)
            if not data:
                break
            print(f"Mensagem recebida do servidor: {data.decode('utf-8')}")

    except ConnectionRefusedError:
        print("Erro: Não foi possível conectar ao servidor")
    except Exception as e:
        print(f"Erro inesperado: {e}")
    finally:
        # Fechar conexão
        client_socket.close()
        print("Conexão fechada")

if __name__ == "__main__":
    tcp_client()